tools/system_libs.py

```diff
@@ -78,161 +78,480 @@ def create_lib(libname, inputs):
+        return glob_in_path(self.src_dir, self.src_glob, self.src_glob_exclude or ())
+
+    raise NotImplementedError()
+
+  def build_objects(self):
+    """
+    Returns a list of compiled object files for this library.
+
+    By default, this builds all the source files returned by `self.get_files()`,
+    with the `cflags` returned by `self.get_cflags()`.
+    """
+    commands = []
+    objects = []
+    cflags = self.get_cflags()
+    for src in self.get_files():
+      o = self.in_temp(os.path.basename(src) + '.o')
+      commands.append([shared.PYTHON, self.emcc, src, '-o', o] + cflags)
+      objects.append(o)
+    run_commands(commands)
+    return objects
+
+  def build(self):
+    """Builds the library and returns the path to the file."""
+    lib_filename = self.in_temp(self.get_name())
+    create_lib(lib_filename, self.build_objects())
+    return lib_filename
+
+  @classmethod
+  def _inherit_list(cls, attr):
+    # Some properties, like cflags and includes, makes more sense to inherit
+    # via concatenation than replacement.
+    result = []
+    for item in cls.__mro__[::-1]:
+      # Using  __dict__ to avoid inheritance
+      result += item.__dict__.get(attr, [])
+    return result
+
+  def get_cflags(self):
+    """
+    Returns the list of flags to pass to emcc when building this variation
+    of the library.
+
+    Override and add any flags as needed to handle new variations.
+    """
+    cflags = self._inherit_list('cflags')
+    cflags += get_cflags(force_object_files=self.force_object_files)
+
+    if self.includes:
+      cflags += ['-I' + shared.path_from_root(*path) for path in self._inherit_list('includes')]
+
+    return cflags
+
+  def get_base_name_prefix(self):
+    """
+    Returns the base name of the library without any suffixes.
+    """
+    return self.name
+
+  def get_base_name(self):
+    """
+    Returns the base name of the library file.
+
+    This will include suffixes such as -mt, but will not include a file extension.
+    """
+    return self.get_base_name_prefix()
+
+  def get_ext(self):
+    """
+    Return the appropriate file extension for this library.
+    """
+    return 'a' if shared.Settings.WASM_BACKEND and shared.Settings.WASM_OBJECT_FILES else 'bc'
+
+  def get_name(self):
+    """
+    Return the full name of the library file, including the file extension.
+    """
+    return self.get_base_name() + '.' + self.get_ext()
+
+  def get_symbols(self):
+    """
+    Return a list of symbols this library exports. See `symbols`.
+
+    This is the dynamic version of `symbols`.
+    """
+    return self.symbols.copy()
+
+  def get_depends(self):
+    """
+    Return a list of simple names of libraries that this library depends on.
+
+    This is the dynamic version of `depends`.
+    """
+    return self.depends
+
+  @classmethod
+  def vary_on(cls):
+    """
+    Returns a list of boolean constructor arguments that defines the variations
+    of this library.
```

Comment:
It would be good to clarify that these are strings. When I initially read this I thought "boolean contructor arguments" meant bools.